airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,257 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+import os
+import urllib
+from pathlib import Path
+from typing import Any, Mapping, Optional, Tuple, Union
+
+import requests
+import requests_cache
+from airbyte_cdk.models import FailureType
+from airbyte_cdk.sources.http_config import MAX_CONNECTION_POOL_SIZE
+from airbyte_cdk.sources.streams.call_rate import APIBudget, CachedLimiterSession, LimiterSession
+from airbyte_cdk.utils.constants import ENV_REQUEST_CACHE_PATH
+from airbyte_cdk.utils.traced_exception import AirbyteTracedException
+from requests.auth import AuthBase
+
+from .decoders import Decoder, JsonDecoder
+from .error_handlers import DefaultBackoffStrategy, HttpStatusErrorHandler, JsonErrorMessageParser, ResponseAction
+from .exceptions import DefaultBackoffException, RequestBodyException, UserDefinedBackoffException
+from .rate_limiting import default_backoff_handler, user_defined_backoff_handler
+
+BODY_REQUEST_METHODS = ("GET", "POST", "PUT", "PATCH")
+
+
+class HttpClient:
+    def __init__(
+        self,
+        stream_name: str,
+        logger: logging.Logger,
+        api_budget: Optional[APIBudget] = None,
+        session: Optional[Union[requests.Session, requests_cache.CachedSession]] = None,
+        authenticator: Optional[AuthBase] = None,
+    ):
+        self._stream_name = stream_name
+        self._api_budget: APIBudget = api_budget or APIBudget(policies=[])
+        if session:
+            self._session = session
+        else:
+            self._session = self.request_session()
+            self._session.mount(
+                "https://", requests.adapters.HTTPAdapter(pool_connections=MAX_CONNECTION_POOL_SIZE, pool_maxsize=MAX_CONNECTION_POOL_SIZE)
+            )
+        if isinstance(authenticator, AuthBase):
+            self._session.auth = authenticator
+        self._logger = logger
+        self._error_handler = HttpStatusErrorHandler(logger)
```

Comment:
_error_handler

Reply:
Agreed. these are good defaults but they all need to be injected

Reply:
How would y'all f eel about keeping the `HttpStatusErrorHandler` abstracted away from connectors, but we allow for some sort of structured map which we use to instantiate on the `HttpStatusErrorHandler`? We lose some flexibility, but if we really just care about error code / response text -> airbyte error type, we've gotten burned in the past w/ some of these fully customizable interface overrides.

But regardless of the choice we make, the injectable object will be optional so surfacing the parameter would come in the follow up PR to make the mappings configurable